data_safe_haven/commands/shm.py

```diff
@@ -4,114 +4,144 @@
 
 import typer
 
-from data_safe_haven.config import DSHPulumiConfig, SHMConfig
-from data_safe_haven.context import ContextSettings
-from data_safe_haven.exceptions import DataSafeHavenError, DataSafeHavenInputError
-from data_safe_haven.external import GraphApi
-from data_safe_haven.infrastructure import SHMProjectManager
+from data_safe_haven import console
+from data_safe_haven.config import ContextManager, SHMConfig
+from data_safe_haven.exceptions import (
+    DataSafeHavenAzureAPIAuthenticationError,
+    DataSafeHavenAzureError,
+    DataSafeHavenConfigError,
+)
+from data_safe_haven.infrastructure import ImperativeSHM
+from data_safe_haven.logging import get_logger
+from data_safe_haven.validators import (
+    typer_aad_guid,
+    typer_fqdn,
+)
 
 shm_command_group = typer.Typer()
 
 
 @shm_command_group.command()
 def deploy(
```

Comment:
This command is doing a lot of work,

- Create or fetch SHM config
- Update SHM config with write protection
- Deploy SHM

It breaks the convention we had before of `context add/config upload` and `deploy`.

Do I understand correctly that before `dsh shm deploy` there is no where to store config?
How likely is it that someone will want to modify an already deployed SHM?

Reply:
> we could do something like
> ```
> dsh config shm-template
> # edit template
> dsh shm deploy shm.yaml
> ```

The main issue with this is that SHMConfig now has 6 fields, 4 of which we don't want the user to edit:

```yaml
azure:
  location: uksouth  # user-editable
  subscription_id: <GUID from lookup>
  tenant_id: <GUID from lookup>
shm:
  admin_group_id: <GUID from lookup>
  entra_tenant_id: <GUID from lookup>
  fqdn: green.develop.turingsafehaven.ac.uk  # user-editable
```

I'm not sure how much the user will benefit from editing these two fields in a config file (while remembering not to edit the other ones) rather than providing them at the command line.

If you feel strongly that this is better, I'm happy to implement it.

> As I understand it, until you deploy there isn't anywhere to upload the config to anyway.

Yes, that's right. What happens at present is that the `SHMConfig` is held in-memory until the storage account has been deployed and is then uploaded.